### ISSUE TITLE
ICLR 26 | TROLL: Trust Regions Improve RL for Large Language Models

### ISSUE THREAD

**Issue by hiahei** [AI-Infra-Team/awesome-papers#169]
# TROLL: Trust Regions Improve RL for Large Language Models
Authors: Philipp Becker, Niklas Freymuth, Serge Thilges, Fabian Otto, Gerhard Neumann
[Paper](https://arxiv.org/abs/2510.03817)

## Motivation

Reinforcement Learning with PPO-like **clipping objectives** has become the standard for fine-tuning Large Language Models. However, clipping suffers from several critical issues:

- **Theoretical Gap**: Clipping is a crude, heuristic approximation of principled KL-based trust regions.
- **Optimization Instability**: It often leads to unstable updates, poorly calibrated gradients, and high sensitivity to hyperparameters.
- **Vanishing Gradients**: Clipping suppresses gradients entirely for tokens where the importance ratio exceeds the threshold, hindering efficient learning.

## Method

**Trust Region Optimization for Large Language models (TROLL)** replaces the clipping mechanism with a **discrete differentiable trust region projection**.

1. **Differentiable Projection**: TROLL formulates a convex optimization problem that projects the new policy distribution onto a KL-trust region around the old policy.

2. **Sparsification Scheme**: To handle massive LLM vocabularies (e.g., 150k+ tokens), TROLL operates on a sparse subset of the most important token logits (typically the top 5-10 tokens), preserving >99.999% of the probability mass.

3. **OptNet Framework**: It utilizes implicit differentiation to maintain gradient information even for updates constrained by the trust region.

4. **Drop-in Replacement**: It acts as a direct replacement for PPO-style clipping during training and requires **zero overhead** during model inference.

<img width="972" height="149" alt="Image" src="https://github.com/user-attachments/assets/c11ffba3-cd5f-4c29-a2ee-082da6eda39b" />

## Experiment

<img width="842" height="465" alt="Image" src="https://github.com/user-attachments/assets/21b91458-a4a0-4072-bc67-b16f4630dee0" />

<img width="850" height="204" alt="Image" src="https://github.com/user-attachments/assets/9cdf2fcb-f9fc-463f-afa2-e75b57043941" />

<img width="847" height="384" alt="Image" src="https://github.com/user-attachments/assets/451b0617-ea18-41fe-952c-754ba858af43" />
